### PR TITLE
Apache Pulsar - pubsub component

### PR DESCRIPTION
**Pull request by elyst**
# Description

This pull request extends the pubsub library of Dapr by adding Apache Pulsar.
 
## Issue reference

#345 

## Checklist

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_600_

### CODE REVIEW

**Comment by yaron2** [comment]
Thanks for this PR.

Could you please make sure that we use Pulsars failover mode for subscriptions?

http://pulsar.apache.org/docs/en/concepts-messaging/#failover

**Comment by elyst** [comment]
> Thanks for this PR.
> 
> Could you please make sure that we use Pulsars failover mode for subscriptions?
> 
> http://pulsar.apache.org/docs/en/concepts-messaging/#failover

Thanks for the feedback.
The subscription type has been set to Failover

File: pubsub/pulsar/pulsar.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+func parsePulsarMetadata(meta pubsub.Metadata) (*pulsarMetadata, error) {
+	m := pulsarMetadata{}
+
+	if val, ok := meta.Properties[host]; ok && val != "" {
+		m.Host = val
+	} else {
+		return nil, errors.New("pulsar error: missing pulsar host")
+	}
+	if val, ok := meta.Properties[subscriptionName]; ok && val != "" {
```
Comment (yaron2) [inline pubsub/pulsar/pulsar.go:40]: The Dapr runtime will automatically populate the metadata with a `consumerID` field. This should be used for the subscription name, so you can remove the subscription name from metadata.

See this example for the Kafka Pub/Sub:
https://github.com/dapr/components-contrib/blob/8eff31b19159e3438df98f5c60f6fe4c63c14527/pubsub/kafka/kafka.go#L229

File: pubsub/pulsar/pulsar.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+func (p *Pulsar) HandleMessage(m pulsar.ConsumerMessage, topic string, handler func(msg *pubsub.NewMessage) error) {
+	err := handler(&pubsub.NewMessage{
+		Data:  m.Payload(),
+		Topic: topic,
+	})
+	if err != nil {
+		p.logger.Debugf("Could not handle topic %s", topic)
+	}
+	m.Ack(m.Message)
```
Comment (yaron2) [inline pubsub/pulsar/pulsar.go:126]: The message should be acknowledged only if there's no error returned from the handler.

**Comment by yaron2** [comment]
Thanks for your contribution and welcome to the Dapr community

**Pull request merged**